### PR TITLE
Refactor isAdPositionFixed into ad-helper.js

### PR DESCRIPTION
**Pull request by muxin**
#4122

### CODE REVIEW

**Comment by zhouyx** [comment]
Is `ad-helper.js` going to serve as a helper function file to both a4a and 3p ad? Why can't we put the function to `amp-ad.js` file or `amp-ad-api-handler.js` file?

File: src/ad-helper.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * @param {!Window} win
+ * @return {boolean} whether element or its ancestors have position
+ * fixed (unless they are POSITION_FIXED_TAG_WHITELIST).
+ * This should only be called when a layout on the page was just forced
+ * anyway.
+ */
+export function isAdPositionFixed(el, win) {
+  let hasFixedAncestor = false;
+  do {
+    if (POSITION_FIXED_TAG_WHITELIST[el.tagName]) {
```
Comment (jridgewell) [inline src/ad-helper.js]: We need to pull `POSITION_FIXED_TAG_WHITELIST` in, too.

**Comment by muxin** [comment]
@zhouyx I want to use this method in amp-iframe but amp-iframe cannot have dependency on amp-ad.

**Comment by zhouyx** [comment]
I see, the file name confused me.
As @jridgewell points out you may also need to move `POSITION_FIXED_TAG_WHITELIST`. But I do believe `POSITION_FIXED_TAG_WHITELIST` is a specially list for ad, I can't think of any `amp-iframe` inside `amp-sticky-ad` or `amp-fx-flying-carpet`.

**Comment by muxin** [comment]
@zhouyx The method `isAdPositionFixed` returns true if we allow the element to render, returns false if we doesn't allow it to render(if it's position fixed without whitelisted ancestor). `amp-iframe` can be inside `amp-light-box`, we will allow it to render even if `amp-light-box` is position: fixed. `amp-iframe` inside `amp-sticky-ad` is not possible but it doesn't hurt to check anyway.

**Comment by zhouyx** [comment]
My question is `POSITION_FIXED_TAG_WHITELIST` is something used only for ad right now and I think you are adding `amp-light-box` to it for `amp-iframe`. Would it be the case that we need different lists for `amp-iframe` and `amp-ad` in future?

**Comment by muxin** [comment]
I think we basically treat `amp-iframe` and `amp-ad` the same regarding the position fixed issue.

**Comment by zhouyx** [comment]
Sure, I am good with it, we have validation check anyway.

**Comment by dvoytenko** [comment]
@muxin @zhouyx to confirm: amp-ad and amp-iframe only treated the same w.r.t. to positioning when we believe a particular instance of an amp-iframe is essentially an ad.

**Comment by muxin** [comment]
@dvoytenko PTAL

**Comment by zhouyx** [comment]
Ah, if we believe a particular instance of an amp-iframe is an ad, then it do makes sense to refactor the function. @dvoytenko What are cases when we believe an iframe is actually an ad?

**Comment by muxin** [comment]
@zhouyx https://github.com/ampproject/amphtml/blob/master/extensions/amp-iframe/0.1/amp-iframe.js#L479

File: src/ad-helper.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * @param {!Element} el
+ * @param {!Window} win
+ * @return {boolean} whether element or its ancestors have position
+ * fixed (unless they are POSITION_FIXED_TAG_WHITELIST).
+ * This should only be called when a layout on the page was just forced
+ * anyway.
+ */
+export function isAdPositionFixed(el, win) {
```
Comment (dvoytenko) [inline src/ad-helper.js]: Now, looking at this method and its uses, I definitely believe that we need to reverse true and false here and call it `isAdPositionAllowed`.

File: src/ad-helper.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** @const These tags are allowed to have fixed positioning */
+const POSITION_FIXED_TAG_WHITELIST = {
```
Comment (dvoytenko) [inline src/ad-helper.js]: This module needs tests as well.

File: src/ad-helper.js (hunk 2 above)
Comment (muxin) [inline src/ad-helper.js]: `isAdPositionAllowed` or `isAdPositionFixedAllowed`?

Comment (dvoytenko) [inline src/ad-helper.js]: isAdPositionAllowed

Comment (muxin) [inline src/ad-helper.js]: done

File: src/ad-helper.js (hunk 3 above)
Comment (muxin) [inline src/ad-helper.js]: done

**Comment by dvoytenko** [comment]
LGTM

**Pull request merged**